### ISSUE TITLE
Parameterization of storeId and artifactId in templates and pipelines

### ISSUE THREAD

**Issue by varshavaradarajan** [gocd/gocd#4206]
##### Issue Type
- Feature enhancement

##### Summary

Pluggable artifact configuration can be configured at a job level in templates and pipelines. When using this configuration with templates, this configuration can be parameterized. Specifically, the store id. Given that the store id can be parameterized, different pipelines using the templates can define different store ids for stores from different plugins. So, essentially, for a job in the template, the specific plugin properties that need to be configured will be based on what the store id is. This creates a problem because users may end up specifying properties that are not useful for a plugin (will result in validation errors on all the pipelines, when configured through the UI and API). Moreover, for the view, we cannot even determine what the plugin view for configuration should be if we cannot determine what the plugin id is from the store id, as it is parameterized.

Similarly, for fetch pluggable artifact, since the artifactId can be parameterized, plugin specific properties could potentially result in not being useful and the fetch artifact view cannot be determined.  

We decided to work around this by adding the plugin id attribute on the pluggable artifact configuration and the fetch pluggable artifact configuration. 

- pluginID attribute cannot be parameterized. 
- Validate whether the plugin id associated with the store id or artifact id is the same as the plugin id configured by the user. 

Here's how the configuration will look like - 

```json
<pluggableArtifact id="demo-image" storeId="#{param}" pluginId="cd.go.artifact.docker">
  <property>
    <key>BuildFile</key>
    <value>image_info.json</value>
  </property>
</pluggableArtifact>
```
```json
<fetchPluggableArtifact artifactId="#{another-param}" pluginId="cd.go.artifact.docker" pipeline="#{pipeline}" stage="build" job="build_docker_image">
  <property>
    <key>Destination</key>
    <value>somewhere-on-the-agent</value>
  </property>
</fetchPluggableArtifact>
```

cc @arvindsv @bdpiparva

**Comment by arvindsv** [comment]
The fact that `storeId` is parameterizable is the one causing the problem here, right? If `storeId` is parameterizable, that tells users that stores can change across pipelines. That's not correct, given that the params will be different per store (and hence plugin).

So, disallowing that, will make all of this work well. Right? This is my ideal:

```xml
<artifactStore pluginId="plugin.1" id="cannot-be-param">
...
</artifactStore>
```

```xml
<artifact id="demo-image" storeId="cannot-be-param">
  <property>
    <key>BuildFile</key>
    <value>image_info.json</value>
  </property>
</artifact>
```

```xml
<fetchartifact artifactId="#{can-be-param}" pipeline="#{can-be-param-pipeline}" stage="build" job="build_docker_image">
  <property>
    <key>Destination</key>
    <value>somewhere-on-the-agent</value>
  </property>
</fetchartifact>
```

Notice that I strongly dislike `fetchPluggableArtifact` and `pluggableArtifact`. I think giving users the information that they're pluggable is unnecessary and incorrect.

**Comment by arvindsv** [comment]
We should also start considering failing things at runtime. I know it'll complicate UI views which need to deal with incorrect data, if they talk to a plugin - but we will never be able to get rid of that. Imagine someone editing config XML directly and making the params invalid. That'll be the same situation anyway.

**Comment by arvindsv** [comment]
If you really want `storeId` to be parameterizable, then yes, you'll probably want to have plugin ID duplicated everywhere. :(

It feels icky and you will still not be solving the problem of someone directly editing the config XML. So, even if you put the plugin ID everywhere, make sure that the case of direct edit is still handled gracefully.

**Comment by arvindsv** [comment]
I can see why it will be ideal to have the `storeId` be parameterized. If you're using one store of type pluginX for one pipeline, and another store of type pluginX for another pipeline. So, my earlier point about not making the `storeId` configurable is wrong, assuming we want to provide that flexibility.

**Comment by arvindsv** [comment]
@varshavaradarajan Yes, that encapsulates what we talked about. One thing I didn't mention is: Just because a template is being used doesn't mean that the plugin ID cannot be determined. Maybe the template is used for templatizing something else.

**Comment by varshavaradarajan** [comment]
@arvindsv - yes. That's why I mentioned "when plugin id cannot be determined". The plugin view can be available at the template level as well.

**Comment by jyotisingh** [comment]
I don't think I get enough context and clarity from this [comment](https://github.com/gocd/gocd/issues/4206#issuecomment-387156213) to be able to say much. I am sure in your heads its all figured out :)

About `As far as possible, we should try to determine the plugin in order to show the view`:
At a pipeline level configuration (with no templates), this is never going to be an issue - parameter or not. 
The only thing that we need to think about is while editing templates with associated pipelines. 
Again, as long as `storeId` is not a parameter - we have nothing to be worried about.
Problem arises when storeId is parameterized. I gather from your comment that we should determine pluginid in all cases apart from the last one wherein storeId is parameterized. So far so good. 

About the key-value pair thing that you talk about which would come into picture for the scenario where storeId is parameterized. Would this just have a free form key value pair that would not be validated at the time of config-save but would be passed on to the plugin in context at the time of execution? With this there is a chance that plugins could receive a list of properties that it doesn't know about or has mandatory fields missing etc. Such cases need to be handled gracefully during execution. There might be things to be taken care of from implementation point of view, but I dont really see any concerns with this approach. I think it makes sense.

**Comment by varshavaradarajan** [comment]
> About the key-value pair thing that you talk about which would come into picture for the scenario where storeId is parameterized. Would this just have a free form key value pair that would not be validated at the time of config-save but would be passed on to the plugin in context at the time of execution?

This form will be at the template level.  While saving the templates, even the pipelines associated with the template are validated. At this point, using the information entered in the key value pair form, plugin validations can happen. If one of the pipeline validations fail, the template will not be allowed to save. 

>  With this there is a chance that plugins could receive a list of properties that it doesn't know about . 

Yes. I could use the same template to provide values for 2 different plugins if the storeId is a parameter. Each plugin will receive all properties and validate whether the properties that it needs are available and meets other constraints. If not, the save fails.

**Comment by jyotisingh** [comment]
Ah I see. So plugin validations in context of each pipeline associated with the template would still happen if template was saved from UI/API. But if someone messed around with config.xml itself, then we let the user know about the error at the time of job execution, correct? If so, sounds alright to me.

**Comment by varshavaradarajan** [comment]
Yes. I think that what we do in case of pluggable tasks is the same that we can follow here for external artifacts. If edited through admin/config_xml, none of the plugin validations happen. (I know we have spoken about this before) So, we continue with the job as is. If say, the image and tag (taking docker registry artifact plugin as an example) specified in the config are available on the agent, then we continue as is. Otherwise, fail with a meaningful error message. 

Digressing, the reason why none of the plugin validations happen when the config is edited through `admin/config_xml` is because it is the same config save flow used during the server startup. If any plugin was missing at the time, the config save would have failed and the previous version of the config would have been picked up from `config.git` for validation. This would have continued until a config xml which didn't contain the plugin's configuration was picked up, config save validation became successful and the server started. This meant that a lot of legitimate config changes can be lost. I think that we can distinguish between a config save at startup and a config save through `admin/config_xml` using some flag and validate the plugin related configs. But, I think you had something against this, @jyotisingh ? Can't remember what it is right now. Do you?

**Comment by varshavaradarajan** [comment]
Ah yes. I'm sure that was @jyotisingh's reason too (surprise, surprise). So, we don't want to do plugin validations for full config save ever?

**Comment by varshavaradarajan** [comment]
Also, when config save happens through the UI (the old one), it's still a full config save, right? We're okay with plugin validations there.

**Comment by arvindsv** [comment]
It does use that full config save endpoint. But it's possible to know which part changed, from the Rails controller side.

**Comment by jyotisingh** [comment]
> So, we don't want to do plugin validations for full config save ever?

Yes, if by full config save you mean when user submits the "full xml" - from api, ui or changes from filesystem. Hence the need for meaningful messages at the execution time incase something fails. 

> Also, when config save happens through the UI (the old one), it's still a full config save, right? We're okay with plugin validations there.

I assume you are talking about the pipeline config save using old UI. We already have plugin validations in that flow (for the things that have existed so far). But ofcourse, since the old UI had actions dealing with only a subset of the entity, the invocations are spread across - such as [this](https://github.com/gocd/gocd/blob/1fb2812cfa45eee8502e420443f2b46d81c51f98/server/webapp/WEB-INF/rails.new/app/controllers/admin/tasks_controller.rb#L75) or [this](https://github.com/gocd/gocd/blob/1fb2812cfa45eee8502e420443f2b46d81c51f98/server/webapp/WEB-INF/rails.new/app/controllers/admin/materials/pluggable_scm_controller.rb#L167).

**Comment by bdpiparva** [comment]
Just to summarize the conversation,

- Creating a new template:
  * Pluggable artifact: 
    > StoreId is not parameterized

    In this scenario storeId is not parameterized hence we can determine the plugin id and view for the pluggable artifact.

     > Store id is parameterized

     In this case, plugin id is not available while configuring the template so provide a form where the user can specify artifact configuration using key-value pairs - No plugin validation at this point

     Alternate option: show plugin dropdown and on select of a plugin, load the pluggable artifact view. No plugin validation at this point. This will make sure that user is configuring right values.

  * Fetch artifact:
     > Artifact id is not parameterized

     Determine plugin id based on artifact id and load the fetch artifact view

     > Artifact id is parameterized

      Again provide a key-value pair to configure defaults - No plugin validation at this point

      Alternate option: show plugin dropdown and on select of a plugin, load the fetch artifact view.

- Creating new pipeline from the template:
  - Specify all required parameter(pipeline cannot be created when a template has parameterized store id or artifactId but pipeline don’t specify a value for it)
  - Validate config on save

**Comment by varshavaradarajan** [comment]
@bdpiparva - Yes. Seems right. I'm still doodling on paper about what errors to show, what happens when editing the template or pipeline, if the value for the parameter used for storeid/artifactid changes.

**Comment by akshaydewan** [comment]
> Alternate option: show plugin dropdown and on select of a plugin, load the pluggable artifact view. No plugin validation at this point. This will make sure that user is configuring right values.

The view can look something like the Materials view, wherein we allow multiple pluginIds to be selected and added:

![screen shot 2018-05-09 at 10 59 47](https://user-images.githubusercontent.com/6024038/39797148-29c7564e-5378-11e8-8d68-0c7261e4af88.png)

@naveenbhaskar @tinozza - Your inputs will be needed here

**Comment by varshavaradarajan** [comment]
To elaborate on what @akshaydewan and @bdpiparva are suggesting - 

If the plugin id cannot be determined, then instead of the previously suggested key-value pair form + plugin lookup, provide an option to select multiple plugins on the UI, provide the multiple plugin views, configure the values for each of them.  Upon clicking save, they values entered will still be provided as a key value pair in the request, but providing the plugin views will make the UX better.  

Concern: Its going to be a huge form for configuring publish artifact and fetch artifact if we have to show views for multiple plugins.

**Comment by jyotisingh** [comment]
@bdpiparva - In [this](https://github.com/gocd/gocd/issues/4206#issuecomment-387617349) comment, you state:
<img width="647" alt="screen shot 2018-05-09 at 11 27 00 am" src="https://user-images.githubusercontent.com/1046342/39799325-8423d960-5381-11e8-898c-bc2a60fbaae3.png">


Since the artifactId is not unique across the system, you would need pipeline::stage::job and the artifactId to determine the pluginId. Even if one of those is parameterized, you would have to use option 2 ie. free form key-value pairs or something else that you guys come up with.

**Comment by varshavaradarajan** [comment]
Scenarios to take care of -

The xml representation -

```
<artifact type="external" id="artifact1" storeId="some_global_store">
  <configuration>
    <property>
      <key>Property1</key>
      <value>Value1</value>
    </property>
     <property>
      <key>Property2</key>
      <value>#{param}</value>
    </property>
    <property>
      <key>Property3</key>
      <encryptedValue>encryptedvalue</encryptedValue>
    </property>
  </configuration>
</artifact>
```
Everything under the `configuration` tag comes from the plugin when configuring from the UI.

**Pipeline Level** - ***Publish Artifact Store Config***

1. Non parameterized store id - may be a dropdown could be given to choose the global store id?
2. Store id is a parameter and the parameter is not yet defined at the pipeline level.
3. Store id is a parameter and the parameter value configured is not one of the global artifact stores' id.
4. Store id is a parameter and a valid parameter has been defined. 
5. If the parameter value changes after the publish artifact store config is saved?

**Template Level** - ***Publish Artifact Store Config***

1. Non parameterized store id. Then the view from the plugin should be shown.
2. Store id is a parameter. Multiple plugin related properties can be configured. 
3. If store id is not a valid parameter in one of the associated pipelines, saving the config should fail. If the template is being created, then there are no  associated pipelines, so the save will be successful. 

@akshaydewan , @jyotisingh - Am I missing anything else?

cc @naveenbhaskar , @tinozza

**Comment by sheroy** [comment]
I'd like to revisit some decisions made earlier in this thread. I have similar concerns as @arvindsv about exposing a selectable Plugin ID across plugin configuration screens. I couldn't tell why we think we can't make store id non-parametrizable at a template level. I think including the plugin Id dropdown on a fetch artifact configuration screen adds an extra level of complexity that we can do without. 

@akshaydewan @varshavaradarajan @arvindsv @ketan

**Comment by varshavaradarajan** [comment]
@sheroy - I should have updated this ticket as well after the discussion and standup. Sorry. Let's start with non parameterizable store id (publish) and artifact id (fetch) and why I don't think we should do that:

- For the fetch external artifact configuration, pipeline, stage, job and artifact id is required to identify which upstream artifact to fetch. All these four inputs are required from the user. Today, we allow parameterizing for pipeline, stage and job for fetch artifact tasks. So, disallowing parameterization of just the artifact id won't be useful. 

- A template is basically just a pipeline configuration that can be reusable. I don't think we should restrict this simply because _sometimes_ users may have to select a plugin id. The UX needs this additional step to choose a plugin id only when it can't be detected by the store and artifact id. If you think that users may not want to ever parameterize the store and artifact id, then the ux for that is just the same as everywhere else when a form from a plugin needs to be shown. 

- The UI is not the only way for people to configure the pipelines. There's xml, json, yaml, apis. I don't think we should restrict the templates and make the store and artifact id non parameterizable simply because sometimes some users who like to use the UI may have to  go through the additional trouble of selecting a plugin id from a dropdown. 

Next, regarding the views themselves(https://github.com/gocd/gocd/issues/4752 and https://github.com/gocd/gocd/issues/4767) - I think that https://github.com/gocd/gocd/issues/4752#issuecomment-391256951 sort of implies that users have to select the plugin id all the time. That's not the case. It is only when the store id entered by the user does not help in determining which plugin the user wants the form for.  

Let me know what you think. 

@arvindsv, @akshaydewan - since we are working on the views already, it would be best if you could reply to this soon so that we are all on the same page.

**Comment by arvindsv** [comment]
@varshavaradarajan We can all talk once you're back tomorrow. /cc @sheroy
